### PR TITLE
Refactor ASN1/x509/PKCS7/ProtoBuf

### PR DESCRIPTION
**Pull request by wargio**
# do not squash

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

I have refactored and cleaned up the ASN1/x509/PKCS7/ProtoBuf code to follow the coding guidelines.
Most of the changes has been performed with search & replace
I have added also tests since they were missing

### CODE REVIEW

File: librz/core/cmd/cmd_print.c [hunk 1]
```diff
@@ -1073,13 +1073,12 @@ static void print_format_help_help_help_help(RzCore *core) {
...
-		RASN1Object *asn1 = rz_asn1_create_object(data, size);
+		RzASN1Object *asn1 = rz_asn1_object_parse(data, size);
 		if (asn1) {
-			char *res = rz_asn1_to_string(asn1, 0, NULL);
-			rz_asn1_free_object(asn1);
+			char *res = rz_asn1_to_string2(asn1, 0, input[1] != 'q');
+			rz_asn1_object_free(asn1);
 			if (res) {
-				rz_cons_printf("%s\n", res);
+				rz_cons_printf("%s", res);
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c:1081]: Is this an intended change?

File: librz/core/cmd/cmd_print.c [hunk 2]
```diff
@@ -1088,31 +1087,31 @@ static void cmd_print_fromage(RzCore *core, const char *input, const ut8 *data,
...
 	} break;
 	case 'p': // "pFp"
 	{
-		RCMS *cms = rz_pkcs7_parse_cms(data, size);
+		RzCMS *cms = rz_pkcs7_cms_parse(data, size);
 		if (cms) {
 			char *res = rz_pkcs7_cms_to_string(cms);
 			if (res) {
-				rz_cons_printf("%s\n", res);
+				rz_cons_printf("%s", res);
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c:1111]: Same question here

File: librz/util/asn1.c [hunk 3]
```diff
@@ -158,20 +152,20 @@ static RASN1Object *asn1_create_object(const ut8 *buffer, ut64 length, const ut8
 	return object;
 }
 
-RZ_API RZ_OWN RASN1Object *rz_asn1_create_object(RZ_NONNULL const ut8 *buffer, ut32 length) {
```
Comment (XVilka) [inline librz/util/asn1.c:161]: Add Doxygen for these functions while you are on it

File: librz/util/asn1.c [hunk 4]
```diff
@@ -417,46 +404,57 @@ RZ_API char *rz_asn1_to_string(RASN1Object *object, ut32 depth, RzStrBuf *sb) {
...
+				RzASN1Object *obj = object->list.objects[i];
+				rz_asn1_to_string(obj, depth + 1, structured, sb);
 			}
 		}
 	}
-	return root ? rz_strbuf_drain(sb) : NULL;
 }
 
-RZ_API void rz_asn1_free_object(RASN1Object *object) {
+RZ_API RZ_OWN char *rz_asn1_to_string2(RZ_NULLABLE RzASN1Object *object, ut32 depth, bool structured) {
```
Comment (XVilka) [inline librz/util/asn1.c]: Probably better to rename `rz_asn1_to_string()` instead, e.g. `*_to_strbuf()` or something? @ret2libc what do you think?

File: librz/util/asn1.c [hunk 5]
```diff
@@ -417,46 +404,57 @@ RZ_API char *rz_asn1_to_string(RASN1Object *object, ut32 depth, RzStrBuf *sb) {
...
+	RzStrBuf *sb = rz_strbuf_new("");
+	if (!sb) {
+		return NULL;
+	}
+	rz_asn1_to_string(object, depth, structured, sb);
+	return rz_strbuf_drain(sb);
+}
+
+RZ_API void rz_asn1_object_free(RzASN1Object *object) {
 	ut32 i;
```
Comment (XVilka) [inline librz/util/asn1.c]: You can move `ut32 i` after return

File: librz/util/astr.c [hunk 6]
```diff
@@ -61,10 +61,10 @@ RZ_API RASN1String *rz_asn1_stringify_string(const ut8 *buffer, ut32 length) {
 		return NULL;
 	}
 	rz_str_filter(str);
-	return rz_asn1_create_string(str, true, strlen(str));
+	return rz_asn1_string_parse(str, true, strlen(str));
 }
 
-RZ_API RASN1String *rz_asn1_stringify_utctime(const ut8 *buffer, ut32 length) {
+RZ_API RZ_OWN RzASN1String *rz_asn1_stringify_utctime(RZ_NULLABLE const ut8 *buffer, ut32 length) {
```
Comment (XVilka) [inline librz/util/astr.c:64]: Doxygen for these functions while you are on it, please.

**Review by XVilka — CHANGES_REQUESTED** [review]

**Comment by XVilka** [comment]
```
Mismatched function annotation for rz_asn1_object_free at <rizin/librz/include/rz_util/rz_asn1.h:101:13> : RZ_NULLABLE (was None at <rizin/librz/util/asn1.c:457:13>
Mismatched function annotation for rz_asn1_binary_free at <rizin/librz/include/rz_util/rz_asn1.h:105:13> : RZ_NULLABLE (was None at <rizin/librz/util/asn1.c:475:13>
Mismatched function annotation for rz_asn1_string_free at <rizin/librz/include/rz_util/rz_asn1.h:104:13> : RZ_NULLABLE (was None at <rizin/librz/util/astr.c:313:13>
```

File: librz/core/cmd/cmd_print.c (hunk 1 above)
Comment (wargio) [reply librz/core/cmd/cmd_print.c:1081]: yes, before it was adding an unnecessary newline

File: librz/core/cmd/cmd_print.c (hunk 2 above)
Comment (wargio) [reply librz/core/cmd/cmd_print.c:1111]: yes, before it was adding an unnecessary newline

**Review by ret2libc — APPROVED** [review]

**Review by XVilka — APPROVED** [review]

**Pull request merged**